source/usr/lib/mcvirt/virtual_machine/virtual_machine.py

```diff
@@ -1230,6 +1231,68 @@ def getVncPort(self):
         else:
             return domain_xml.find('./devices/graphics[@type="vnc"]').get('port')
 
+    def _set_vnc_listen_address(self, domain_xml, address):
+        """Sets the VNC listen address, then generates and returns new random password for VNC."""
+        # Check the user has permission to view the VM console
+        self._get_registered_object('auth').assert_permission(
+            PERMISSIONS.VIEW_VNC_CONSOLE,
+            self
+        )
+
+        vnc_el = domain_xml.find('./devices/graphics[@type="vnc"]')
+        if vnc_el is None:
+            raise VncNotEnabledException('VNC is not enabled on the VM')
+        vnc_el.set('listen', address)
+        listen_el = vnc_el.find('listen[@type="address"]')
+        if listen_el is not None:
+            listen_el.set('address', address)
+        else:
+            listen_el = EL.Element('listen')
+            listen_el.set('type', 'address')
+            listen_el.set('address', address)
+            vnc_el.append(listen_el)
+        passwd = UserBase.generate_password(24)
+        vnc_el.set('passwd', passwd)
+        return passwd
+
+    @Pyro4.expose()
+    @locking_method()
+    def connect_vnc(self):
+        """Returns the hostname, port, and randomly generated password used for the VNC display."""
+        # Check the user has permission to view the VM console
+        self._get_registered_object('auth').assert_permission(
```

Comment:
You will have issues if the VM is not registered on the local node.
Take a look at the 'update_graphics_driver' method for an example of how to overcome this :)